File: tools/server_observer_cpp/src/static_map_cache.cpp
#include "static_map_cache.hpp"
#include <fstream>
#include <filesystem>
#include <iostream>
#include <zstd.h>

namespace fs = std::filesystem;

StaticMapCache::StaticMapCache(const std::string& cache_dir)
    : cache_dir_(cache_dir)
{
    fs::create_directories(cache_dir_);
    
    // Load existing cached map IDs
    for (const auto& entry : fs::directory_iterator(cache_dir_)) {
        if (entry.is_regular_file()) {
            std::string filename = entry.path().stem().string();
            if (filename.find("map_") == 0) {
                try {
                    int map_id = std::stoi(filename.substr(4));
                    saved_ids_.insert(map_id);
                } catch (const std::exception&) {
                    // Skip files with invalid names
                }
            }
        }
    }
}

bool StaticMapCache::is_cached(int map_id) const {
    std::lock_guard<std::mutex> lock(lock_);
    return saved_ids_.find(map_id) != saved_ids_.end();
}

std::string StaticMapCache::save(int map_id, const json& static_map_data) {
    std::lock_guard<std::mutex> lock(lock_);
    
    std::string filename = "map_" + std::to_string(map_id) + ".bin";
    fs::path path = fs::path(cache_dir_) / filename;
    
    // Check if already cached
    if (saved_ids_.find(map_id) != saved_ids_.end() && fs::exists(path)) {
        return path.string();
    }
    
    try {
        // Serialize JSON to string
        std::string json_str = static_map_data.dump();
        
        // Compress using zstd
        size_t compressed_size = ZSTD_compressBound(json_str.size());
        std::vector<uint8_t> compressed(compressed_size);
        
        size_t actual_size = ZSTD_compress(
            compressed.data(), compressed_size,
            json_str.data(), json_str.size(),
            3  // compression level
        );
        
        if (ZSTD_isError(actual_size)) {
            std::cerr << "Compression failed: " << ZSTD_getErrorName(actual_size) << std::endl;
            return "";
        }
        
        compressed.resize(actual_size);
        
        // Write to file
        std::ofstream file(path, std::ios::binary);
        if (!file.is_open()) {
            std::cerr << "Failed to open file for writing: " << path << std::endl;
            return "";
        }
        
        file.write(reinterpret_cast<const char*>(compressed.data()), compressed.size());
        file.close();
        
        saved_ids_.insert(map_id);
        std::cout << "Cached static map data for map_id " << map_id << " at " << path << std::endl;
        
        return path.string();
    } catch (const std::exception& e) {
        std::cerr << "Failed to cache static map data for map_id " << map_id << ": " << e.what() << std::endl;
        return "";
    }
}

File: tools/server_observer/static_map_cache.py
from __future__ import annotations

import pickle
import threading
from pathlib import Path
from typing import Optional
from typing import Set

import zstandard as zstd

from tools.server_observer.recorder_logger import get_logger

logger = get_logger()


class StaticMapCache:
    """
    Cache static map data per map_id to avoid duplicate downloads.
    """

    def __init__(self, cache_dir: Path):
        self.cache_dir = Path(cache_dir)
        self.cache_dir.mkdir(parents=True, exist_ok=True)
        self._saved_ids: Set[int] = set()
        self._lock = threading.Lock()
        self._compressor = zstd.ZstdCompressor(level=3)

        for file in self.cache_dir.glob("map_*.bin"):
            try:
                map_id = int(file.stem.replace("map_", ""))
                self._saved_ids.add(map_id)
            except ValueError:
                continue

    def is_cached(self, map_id: int) -> bool:
        return map_id in self._saved_ids

    def save(self, map_id: Optional[int], static_map_data) -> Optional[Path]:
        if map_id is None or static_map_data is None:
            return None
        with self._lock:
            path = self.cache_dir / f"map_{map_id}.bin"
            if map_id in self._saved_ids and path.exists():
                return path
            try:
                data = pickle.dumps(static_map_data)
                compressed = self._compressor.compress(data)
                with path.open("wb") as f:
                    f.write(compressed)
                self._saved_ids.add(map_id)
                logger.info(f"Cached static map data for map_id {map_id} at {path}")
                return path
            except Exception as exc:
                logger.warning(f"Failed to cache static map data for map_id {map_id}: {exc}")
                return None
